### PR TITLE
chore: Update libraries and use JujuFacade in the charm

### PR DESCRIPTION
**Pull request by saltiyazan**
# Description

This PR:
- Updates the `autounseal`, `vault-kv` and `juju-facade` libraries.
- Uses JujuFacade in the charm and refactors the charm.
- Fixes unit and integration tests.

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/integration/vault_kv_requirer_operator/charmcraft.yaml [hunk 1]
```diff
@@ -18,6 +18,7 @@ storage:
   certs:
     type: filesystem
     minimum-size: 5M
+    location: /var/snap/vault/common/certs
```
Comment (DanielArndt) [inline tests/integration/vault_kv_requirer_operator/charmcraft.yaml:21]: What is the value in setting this?

File: tests/integration/test_integration.py [hunk 2]
```diff
@@ -100,7 +117,10 @@ async def vault_authorized(ops_test: OpsTest, vault_unsealed: Task) -> Task:
     root_token, key = await vault_unsealed
 
     async def authorize():
-        await authorize_charm(ops_test, root_token)
+        try:
+            await authorize_charm(ops_test, root_token)
+        except ActionFailedError:
+            logger.warning("Failed to authorize charm")
```
Comment (DanielArndt) [inline tests/integration/test_integration.py:123]: Won't this have the same effect as before if we just catch the exception and log it?

I would think we would want it to bubble up so the test fails?

(The same question applies to all places we do this in the file)

File: tests/integration/test_integration.py [hunk 3]
```diff
@@ -377,27 +399,29 @@ async def test_deploy_all_the_things(
...
+    await vault_idle_blocked
 
     vault_app = get_app(ops_test.model)
     assert vault_app.status == "blocked"
 
 
 @pytest.mark.abort_on_fail
 async def test_given_certificates_provider_is_related_when_vault_status_checked_then_vault_returns_200_or_429(  # noqa: E501
-    ops_test: OpsTest, vault_idle: Task, self_signed_certificates_idle: Task
+    ops_test: OpsTest, vault_idle_blocked: Task, self_signed_certificates_idle: Task
```
Comment (DanielArndt) [inline tests/integration/test_integration.py:414]: Does this test actually care if vault is blocked?

**Review by DanielArndt — CHANGES_REQUESTED** [review]
There were a couple of tiny changes I just pushed up in a commit so that I could close #339 which was very closely related.

Some other small comments / questions but otherwise LGTM.

File: tests/integration/vault_kv_requirer_operator/charmcraft.yaml (hunk 1 above)
Comment (saltiyazan) [reply tests/integration/vault_kv_requirer_operator/charmcraft.yaml:21]: Nothing, I was getting an error with the `storage-attached` hook, which turned out to be unrelated, but I update this to match what we do in the charm while debugging.

File: tests/integration/test_integration.py (hunk 2 above)
Comment (saltiyazan) [reply tests/integration/test_integration.py:123]: Yes, you are right, nothing will happen now, we could raise the exception, but I didn't want to take action now until we understand what's going on.
Next time that exact intermittent error happens we will have some more info.

File: tests/integration/test_integration.py (hunk 3 above)
Comment (saltiyazan) [reply tests/integration/test_integration.py:414]: No it doesn't care if the charm is blocked or active. However the charm shouldn't be in waiting for example.
`vault_idle` might be true while the leader is still waiting. We might talk to Vault before the snap is installed or started.

**Review by DanielArndt — APPROVED** [review]

File: tests/integration/vault_kv_requirer_operator/charmcraft.yaml (hunk 1 above)
Comment (DanielArndt) [reply tests/integration/vault_kv_requirer_operator/charmcraft.yaml:21]: Im a little anxious to set something explicit if we don't need it, but I don't have a specific reason. I guess if the structure of where snaps store things change, for example.

Comment (saltiyazan) [reply tests/integration/vault_kv_requirer_operator/charmcraft.yaml]: Removed :)

**Pull request merged**